compiler/src/dotty/tools/dotc/core/NameOps.scala

```diff
@@ -231,13 +233,27 @@ object NameOps {
       }
     }
 
-    def functionArity: Int = {
-      def test(prefix: Name): Int =
-        if (name.startsWith(prefix))
-          try name.drop(prefix.length).toString.toInt
-          catch { case ex: NumberFormatException => -1 }
-        else -1
-      test(tpnme.Function) max test(tpnme.ImplicitFunction)
+    def anyFunctionArity: Int =
+      functionArity max implicitFunctionArity max phantomFunctionArity max implicitPhantomFunctionArity
+
+    def functionArity: Int = testArity(tpnme.Function)
+    def implicitFunctionArity: Int = testArity(tpnme.ImplicitFunction)
+    def phantomFunctionArity: Int = testArityWithPhantoms(tpnme.PhantomFunction)
+    def implicitPhantomFunctionArity: Int = testArityWithPhantoms(tpnme.ImplicitPhantomFunction)
+
+    private def testArity(prefix: Name): Int = {
+      if (name.startsWith(prefix))
+        try name.drop(prefix.length).toString.toInt
+        catch { case ex: NumberFormatException => -1 }
+      else -1
+    }
+
+    private def testArityWithPhantoms(prefix: Name): Int = {
+      if (name.startsWith(prefix)) {
+        val phantomicity = name.drop(prefix.length)
+        if (phantomicity.exists(c => c != phantomicityOnChar && c != phantomicityOffChar)) -1
```

Comment:
It is to check if the suffix is valid. Already reformatted and made more explicit.